### PR TITLE
chore(tracing): drops support for v0.3 traces endpoint

### PR DESCRIPTION
**Pull request by mabdinur**
## Description

Currently the python tracer can submit traces to the `v0.3/traces` and `v0.4/tracer` endpoints in the Datadog Agent. The only difference between these endpoints is that `v0.4/traces` returns the agent sample rates (this is "priority sampling") while the response from `v0.3/traces` endpoint does not contain any sampling rates/configurations. 

With the removal of the dd_priority_sampling configuration in a previous PR user's can no longer disable agent based sampling in tracing clients. With this change we should deprecate the `v0.3/traces` endpoint and ensure it is never used. Instead the `v0.4` or `v0.5` endpoints should be used to submit traces to the agent. 

Follow up to: https://github.com/DataDog/dd-trace-py/pull/10102

## Changes 
- Removes `v0.3/traces` endpoint from the tracer AgentWriter. This endpoint is deprecated and will removed from the Agent soon******.
- Removes all tests that submit traces to the `v0.3/traces` agent endpoint
- Renames `v0.3` encoder to `v0.4` encoder (`v0.3/traces` and `v0.4/traces` agent endpoints have the same encoding format)

Note: The minimum supported version  of the datadog agent for ddtrace>=2.0 is [v7.28.0](https://ddtrace.readthedocs.io/en/stable/versioning.html?highlight=supported#release-support). This version of the datadog supports v0.4 and v0.5 trace endpoints. This is not a breaking change. 

## Checklist
- [x] PR author has checked that all the criteria below are met
- The PR description includes an overview of the change
- The PR description articulates the motivation for the change
- The change includes tests OR the PR description describes a testing strategy
- The PR description notes risks associated with the change, if any
- Newly-added code is easy to change
- The change follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html)
- The change includes or references documentation updates if necessary
- Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist
- [x] Reviewer has checked that all the criteria below are met 
- Title is accurate
- All changes are related to the pull request's stated goal
- Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- Testing strategy adequately addresses listed risks
- Newly-added code is easy to change
- Release note makes sense to a user of the library
- If necessary, author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: tests/tracer/test_encoders.py [hunk 1]
```diff
@@ -904,3 +835,12 @@ def test_json_encoder_traces_bytes():
...
     assert "\x80span.b" == span_c["name"]
+
+
+@pytest.mark.subprocess(
+    env={"DD_TRACE_API_VERSION": "v0.3"}, err=".*Using DD_TRACE_API_VERSION=v0.3 is deprecated. Use v0.4 instead.*"
+)
+def test_v03_trace_api_deprecation():
+    from ddtrace import config
+
+    assert config._trace_api == "v0.4"
```
Comment (mabdinur) [inline tests/tracer/test_encoders.py:846]:
```suggestion
     import logging
    from ddtrace import config


    logging.basicConfig()
    assert config._trace_api == "v0.4"
```

File: tests/tracer/test_writer.py [hunk 2]
```diff
@@ -758,11 +757,7 @@ def test_writer_recreate_keeps_headers():
         # Defaults on windows
         ("win32", None, None, False, "v0.4"),
         # Default with priority sampler
-        ("win32", None, None, False, "v0.4"),
-        # Explicitly passed in API version is always used
-        ("win32", "v0.3", None, False, "v0.3"),
-        ("win32", "v0.3", "v0.4", False, "v0.3"),
-        ("win32", "v0.3", "v0.4", False, "v0.3"),
+        ("win32", None, None, "v0.4"),
```
Comment (mabdinur) [inline tests/tracer/test_writer.py:760]:
```suggestion
```

File: tests/tracer/test_writer.py [hunk 3]
```diff
@@ -758,11 +757,6 @@ def test_writer_recreate_keeps_headers():
         # Defaults on windows
         ("win32", None, None, False, "v0.4"),
         # Default with priority sampler
```
Comment (mabdinur) [inline tests/tracer/test_writer.py:759]:
```suggestion
```

File: releasenotes/notes/remove-v03-encoder-1f36d48dc297c6fa.yaml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+---
+deprecations:
+  - |
+    tracing: ``DD_TRACE_API_VERSION=v0.3`` is deprecated. Use ``v0.4`` or ``v0.5`` instead.
```
Comment (mabdinur) [inline releasenotes/notes/remove-v03-encoder-1f36d48dc297c6fa.yaml:4]:
```suggestion
    tracing: ``DD_TRACE_API_VERSION=v0.3`` is deprecated. Use ``v0.4`` or ``v0.5`` instead.

```

File: ddtrace/internal/writer/writer.py [hunk 5]
```diff
@@ -540,38 +539,32 @@ def agent_url(self):
     def _agent_endpoint(self):
         return self._intake_endpoint(client=None)
 
-    def _downgrade(self, payload, response, client):
+    def _downgrade(self, response, client):
```
Comment (mabdinur) [inline ddtrace/internal/writer/writer.py:542]:
```suggestion
    # TODO(munir): v0.5 endpoint was introduced in ddtrace v7.22.0 (August 2020). The vast majority of 
    # users should be using a Datadog Agent that supports the v0.5 endpoint. Soon we should deprecate and
    # remove support for automatically downgrading the trace api. Instead we should ask user's to upgrade
    # their datadog agent or prompt them to set DD_TRACE_API_VERSION to v0.4
    def _downgrade(self, response, client):
```

File: tests/integration/test_integration.py [hunk 6]
```diff
@@ -583,17 +583,16 @@ def encode_traces(self, traces):
...
-
     import mock
 
     from ddtrace import tracer as t
+    from ddtrace.internal.utils.http import Response
 
-    encoding = os.environ["DD_TRACE_API_VERSION"] or "v0.5"
-    t._writer._downgrade(None, None, t._writer._clients[0])
-    assert t._writer._endpoint == {"v0.5": "v0.4/traces", "v0.4": "v0.3/traces"}[encoding]
+    t._writer.api_version = "v0.5"
```
Comment (ZStriker19) [inline tests/integration/test_integration.py:593]: Why this change?

**Review by ZStriker19 — APPROVED** [review]

File: ddtrace/internal/writer/writer.py (hunk 5 above)
Comment (mabdinur) [reply ddtrace/internal/writer/writer.py:542]: This is overkill. If we introduce a new trace api this function could be useful!

File: tests/integration/test_integration.py (hunk 6 above)
Comment (mabdinur) [reply tests/integration/test_integration.py:593]: The downgrade test currently validates updating the trace api from v0.5 to v0.4 and v0.4 to v0,3. Since we are dropping support for v0.3 endpoint we only want to test downgrading the trace api from v0.5 to v0.4.

**Review by erikayasuda — APPROVED** [review]

**Review by emmettbutler — APPROVED** [review]
